discv5/discv5-wire.md

```diff
@@ -1,395 +1,307 @@
+new keys.
 
-- Have the ENR of the recipient include one or more key values (eg:
-  obf1, obf2, etc), and make the *obfuscation parameter* a randomly
-  selected *reference* to one of those.
-- Support for a fully encrypted channel option.
+**TBD: concurrent handshake tie-breaker rule**
 
-### Conversation Nonce
+### Notation
 
-A mandatory component of all messages is the conversation-nonce.
+`[ .. , .. , .. ]`\
+    is recursive encoding as an RLP list\
+`rlp_bytes(x)`\
+    is the RLP encoding of the byte array `x`\
+`a || b`\
+    means binary concatenation of `a` and `b`\
+`xor(a, b)`\
+    means binary XOR of `a` and `b`\
+`sha256(x)`\
+    is the SHA256 digest of `x`\
+`sign(key, x)`\
+    creates a signature of `x` using the given key\
+`aesgcm_encrypt(key, nonce, pt, ad)`\
+    is AES-GCM encryption/authentication with the given `key`, `nonce` and additional\
+    authenticated data `ad`. Size of `key` is 16 bytes (AES-128), size of `nonce` 12 bytes.
 
-A *conversation* is set of message exchanges involving two or more
-participants. A request-reply is a simple conversation. A request for
-nodes or topics followed immediately by a WhoAreYou/IAm is an exchange
-of messages that is also one conversation.
+### Packet Encoding
 
-When a request is sent to a node, the signed, hashed response must
-include some unique code to make sure that the requester can know that
-this is not a replayed message. Another way of viewing this is that a
-correlator must uniquely relate the response with the request.
+All packets start with a fixed-size `tag`. For a packet sent by node A to node B:
 
-This correlator can also serve the purpose of identifying which
-responses correspond to which request if for some reason multiple
-request/replies are happening concurrently.
+    tag              = xor(sha256(dest-node-id), src-node-id)
+    dest-node-id     = 32-byte node ID of B
+    src-node-id      = 32-byte node ID of A
 
-More generally, the conversation nonce groups messages into a unique
-conversation, allowing concurrent conversations to be separated, and
-guaranteeing that replayed messages are ignored.
+The recipient can recover the sender's ID by performing the same calculation in
+reverse.
 
-In future, conversations may involve multiple occurrences of a single
-type of message. An eavesdropper could confuse the conversation by
-replaying previous occurrences. To prevent this, a single byte in the
-conversation nonce will be reserved for future message counting. The top
-4 bytes are the conversation correlator.
+    src-node-id      = xor(sha256(dest-node-id), tag)
 
-```text
-> conversation-nonce= conversation-correlator || reserved-byte
-> conversation-correlator = 4-byte conversation identifier
-> reserved-byte = currently ignored, reserved as a message counter for future versions
-```
+The encoding of the 'random packet', sent if no session keys are available, is:
 
-The value selected must
+    random-packet    = tag || random-data
```

Comment:
After this is merged, I should include a document tying these things to some of the rationale. Eg: one consequence of 'tag' here is that the destination id must be known, which addresses some of the initial requirements / vulnerabilities. I think it's better to keep this relationship between the encoding and the requirements separate from this section, as it needs to be clear and concise, but it would be helpful to add it somewhere.